src/common/cont_props.c

```diff
@@ -33,49 +33,72 @@ daos_props_2cont_props(daos_prop_t *props, struct cont_props *cont_prop)
+		cont_prop->dcp_dedup_size	= daos_cont_prop2dedupsize(props);
 
 	/** checksum */
-	cont_prop->dcp_srv_verify	= daos_cont_prop2serververify(props);
-	cont_prop->dcp_csum_type	= daos_cont_prop2csum(props);
-	cont_prop->dcp_csum_enabled =
-		daos_cont_csum_prop_is_enabled(cont_prop->dcp_csum_type);
-	cont_prop->dcp_chunksize	= daos_cont_prop2chunksize(props);
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_CSUM_SERVER_VERIFY) != NULL)
+		cont_prop->dcp_srv_verify = daos_cont_prop2serververify(props);
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_CSUM)) {
+		cont_prop->dcp_csum_type = daos_cont_prop2csum(props);
+		cont_prop->dcp_csum_enabled =
+			daos_cont_csum_prop_is_enabled(cont_prop->dcp_csum_type);
+	}
+
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_CSUM_CHUNK_SIZE) != NULL)
+		cont_prop->dcp_chunksize = daos_cont_prop2chunksize(props);
 
 	/** compression */
-	cont_prop->dcp_compress_type	= daos_cont_prop2compress(props);
-	cont_prop->dcp_compress_enabled	=
-	       daos_cont_compress_prop_is_enabled(cont_prop->dcp_compress_type);
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_COMPRESS) != NULL) {
+		cont_prop->dcp_compress_type = daos_cont_prop2compress(props);
+		cont_prop->dcp_compress_enabled	=
+			daos_cont_compress_prop_is_enabled(cont_prop->dcp_compress_type);
+	}
 
 	/** encryption */
-	cont_prop->dcp_encrypt_type	= daos_cont_prop2encrypt(props);
-	cont_prop->dcp_encrypt_enabled	=
-		daos_cont_encrypt_prop_is_enabled(cont_prop->dcp_encrypt_type);
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_ENCRYPT) != NULL) {
+		cont_prop->dcp_encrypt_type = daos_cont_prop2encrypt(props);
+		cont_prop->dcp_encrypt_enabled =
+			daos_cont_encrypt_prop_is_enabled(cont_prop->dcp_encrypt_type);
+	}
 
 	/** redundancy */
-	cont_prop->dcp_redun_lvl	= daos_cont_prop2redunlvl(props);
-	cont_prop->dcp_redun_fac	= daos_cont_prop2redunfac(props);
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_REDUN_LVL) != NULL)
+		cont_prop->dcp_redun_lvl = daos_cont_prop2redunlvl(props);
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_REDUN_FAC) != NULL)
+		cont_prop->dcp_redun_fac = daos_cont_prop2redunfac(props);
+
 	/** EC cell size */
-	cont_prop->dcp_ec_cell_sz	= daos_cont_prop2ec_cell_sz(props);
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_EC_CELL_SZ) != NULL)
+		cont_prop->dcp_ec_cell_sz = daos_cont_prop2ec_cell_sz(props);
 
 	/** alloc'ed oid */
-	cont_prop->dcp_alloced_oid	= daos_cont_prop2allocedoid(props);
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_ALLOCED_OID) != NULL)
+		cont_prop->dcp_alloced_oid = daos_cont_prop2allocedoid(props);
 
 	/** performance domain affinity level */
-	cont_prop->dcp_ec_pda		= daos_cont_prop2ec_pda(props);
-	cont_prop->dcp_rp_pda		= daos_cont_prop2rp_pda(props);
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_EC_PDA) != NULL)
+		cont_prop->dcp_ec_pda = daos_cont_prop2ec_pda(props);
+	if (daos_prop_entry_get(props, DAOS_PROP_CO_RP_PDA) != NULL)
+		cont_prop->dcp_rp_pda = daos_cont_prop2rp_pda(props);
```

Comment:
PS leader will retrieve all properties by ALL flags, then IV sync to all engines/targets, so EC/RP_PDA should not be NULL for setting process.  Anyway, this function is only suppose to update cont_prop by props,  it should not reset cont_prop by itself, IMHO.

Reply:
>some case seems not very reliable, like ds_cont_tgt_prop_update() -> cont_child_prop_update() -> daos_props_2cont_props(), for example if the EC/RP_PDA is not included in the prop update, and if don't provide a default value inside daos_props_2cont_props(), then it will not be inited and just be 0, that may cause layout incorrect for PDA (Performanc Domain Affinity) usage.

Right now, the property update (by IV) will always update all container property, i.e. there won't be any individual property update, since DAOS IV cache value can not update single property yet.

TBH, I am not sure where we should add these assertion.  But I think reset these values in this convert function is definitely not right IMHO. 

Though I think this D_DEBUG, you mentioned above, can be used as a warning?

Reply:
“since DAOS IV cache value can not update single property yet”
now the IV_PROP with a "iv_prop->cip_valid_bits" to indicate which set of entries updated, suppose it is intended to supported partial prop update.

If it is partial prop update and missed EC/RP_PDA that may cause layout issue and data corruption, looks we need with assert to check it or fail it when it is not included in that code path?

Reply:
yes, I will update the patch.